lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,276 @@
+  auto executeErr = compiledF.execute(&context);
+  EXIT_ON_ERR(std::move(executeErr));
+  // don't delete bindings.
+  context.movePlaceholderBindings().release();
+}
+
+/// Evaluates a provided constant operation \p C using the provided \p backend
+/// and using the compilation context \p cctx.
+/// \returns constant results.
+std::vector<Constant *>
+evaluateConstantOperation(Backend &backend, CompilationContext &cctx, Node *C) {
+  PlaceholderBindings bindings;
+  assert(isConstantOperation(C, backend) && "Expected a constant expression");
+  // Constants and splats do not need to be constant evaluated.
+  if (isa<Constant>(C) || isa<SplatNode>(C)) {
+    return {};
+  }
+  Module &mod = *C->getParent()->getParent();
+  // Create a temporary function to perform the constant operation.
+  Function *constEvaluationF = mod.createFunction(constEvaluationFunctionName);
+  // Mapping from existing nodes to the new ones.
+  NodeMap currToNew;
+  // Clone the constant operation and some of its inputs if necessary.
+  auto *clonedC = recursiveClone(constEvaluationF, C, currToNew);
+  // Create save nodes for each of the results.
+  llvm::SmallVector<SaveNode *, 16> savedResults;
+  for (unsigned idx = 0, e = clonedC->getNumResults(); idx < e; ++idx) {
+    auto *SN = constEvaluationF->createSave(clonedC->getName(),
+                                            clonedC->getNthResult(idx));
+    savedResults.emplace_back(SN);
+    bindings.allocate(SN->getPlaceholder());
+  }
+  // Run the temporary backend to perform this constant operation
+  // evaluation.
+  EXIT_ON_ERR(
+      executeConstantFunction(backend, *constEvaluationF, bindings, cctx));
+  // Get the results of the constant operation compile-time computation and
+  // create new constants from it.
+  std::vector<Constant *> constResults;
+  constResults.reserve(savedResults.size());
+  for (auto *SN : savedResults) {
+    Tensor *outputTensor = bindings.get(SN->getPlaceholder());
+    auto *constResult = mod.createConstant(SN->getName(), *outputTensor);
+    constResults.emplace_back(constResult);
+  }
+  // Remove the temporary function.
+  mod.eraseFunction(constEvaluationF);
+  return constResults;
+}
+
+/// Check if function \p F consists of constant operations only.
+llvm::Error verifyConstantFunction(Backend &backend, Function &F) {
+  // Perform the checks in DEBUG builds only.
+#ifndef NDEBUG
+  for (auto &N : F.getNodes()) {
+    // Saving results is fine.
+    if (isa<SaveNode>(&N)) {
+      continue;
+    }
+    // Placeholders can be used just to save results.
+    if (!isa<Placeholder>(&N)) {
+      RETURN_ERR_IF_NOT(isConstantOperation(&N, backend),
+                        "Expected constant operation");
+      continue;
+    }
+    if (!N.hasOneUse()) {
+      RETURN_ERR("Expected constant operation");
+    }
+    auto *SN = dyn_cast<SaveNode>(N.getUsers().begin()->getUser());
+    if (SN && SN->getPlaceholder() == &N) {
+      continue;
+    }
```

Comment:
Don't you need to return an error at the end of the for loop here, since it's a PH that isn't used by a Save?

Reply:
Good catch! Thanks!